pauli string reps added (#6243) and bug fixed in pauli arithmetic

ldi18 opened the pull request:
**Context:**
Prior to this commit, PennyLane only allowed access to Pauli string representations for the X, Z and Z gate via `X(0).pauli_rep`, etc. The Pauli representations of other gates were not implemented.

**Description of the Change:**
I added the Pauli string representations for X, Y, Z, H, S, T, SX, SWAP, ISWAP, ECR, SISWAP, RX, RY, RZ, Rot, PhaseShift, U1, U2 and U3. 

The correctness of each Pauli string representation is checked by converting the returned Pauli string into a matrix and comparing it with the respective `.matrix()` method.

The test `test_parametric_ops:TestDecompositions:test_pc_phase_decomposition_broadcasted` (Testing `qml.PCPhase([1.23, 4.56, 7.89], dim=5, wires=[0, 1, 2])`) returned a dimension error. I resolved this by adding an outer product in pauli_arithmetic for tensor-like coefficients (specifically tensor-like phi).

**Benefits:**

Pennylane now supports Pauli string decompositions for various operators. Pauli string decompositions of products of operators with available Pauli string decompositions can now be directly accessed. `Example: (Hadamard(0) @ S(1) @ T(1, 2) @ SX(1) @ SWAP([0, 1])).pauli_rep`

**Possible Drawbacks:**
Classes for parametrized gates may be slightly heavier, as the phase parameter has to be stored. However, the Pauli representation itself is only computed when needed.

**Related GitHub Issues:**
#6243

dwierichs commented on doc/releases/changelog-dev.md:
```diff
@@ -33,6 +33,8 @@
   visualizations, allowing global and per-wire customization with options like `color`, `linestyle`, and `linewidth`.
   [(#6486)](https://github.com/PennyLaneAI/pennylane/pull/6486)
 
+* Added Pauli String representations for the gates X, Y, Z, H, S, T, SX, SWAP, ISWAP, ECR, SISWAP, RX, RY, RZ, Rot, PhaseShift, U1, U2, U2. [(#6243)](https://github.com/PennyLaneAI/pennylane/pull/6562)
```
```suggestion
* Added Pauli String representations for the gates X, Y, Z, H, S, T, SX, SWAP, ISWAP, ECR, SISWAP, RX, RY, RZ, Rot, PhaseShift, U1, U2, U2.
  [(#6562)](https://github.com/PennyLaneAI/pennylane/pull/6562)
```

dwierichs commented on doc/releases/changelog-dev.md:
```diff
@@ -69,13 +71,16 @@
 * Fixed `Identity.__repr__` to return correct wires list.
   [(#6506)](https://github.com/PennyLaneAI/pennylane/pull/6506)
 
+* Fixed a shape error related to tensor-like inputs for the qml.PCPhase gate in pauli_arithmetic. [(#6243)](https://github.com/PennyLaneAI/pennylane/pull/6562)
```
```suggestion
* Fixed a shape error related to tensor-like inputs for the qml.PCPhase gate in pauli_arithmetic.
  [(#6562)](https://github.com/PennyLaneAI/pennylane/pull/6562)
```

dwierichs commented on doc/releases/changelog-dev.md:
```diff
@@ -69,13 +71,16 @@
 Shiwen An
 Astral Cai,
 Yushao Chen,
+Lasse Dierich
```
```suggestion
Lasse Dierich,
```

dwierichs commented on pennylane/ops/qubit/non_parametric_ops.py:
```diff
@@ -77,6 +77,20 @@ def __repr__(self) -> str:
+            )
+        return self._pauli_rep
+
+    def __init__(self, wires: WiresLike, id: Optional[str] = None):
+        super().__init__(wires=wires, id=id)
+
```
Why is this needed?

ldi18 replied:
I added it so that `self._pauli_rep` is explicitly set to `None` (within the super constructor) and because PauliX, PauliY, etc. already had the same call of a super constructor.

I will remove the constructor as there is also no difference in the test cases without it.

dwierichs replied:
Exactly, the super `__init__` is called anyways if there is no new one defined on a subclass :)

dwierichs commented on pennylane/ops/qubit/non_parametric_ops.py:
```diff
@@ -815,6 +850,20 @@ class S(Operation):
+            )
+        return self._pauli_rep
+
+    def __init__(self, wires: WiresLike, id: Optional[str] = None):
+        super().__init__(wires=wires, id=id)
+
```
```suggestion
```

dwierichs commented on pennylane/ops/qubit/non_parametric_ops.py:
```diff
@@ -927,6 +976,20 @@ class T(Operation):
+            )
+        return self._pauli_rep
+
+    def __init__(self, wires: WiresLike, id: Optional[str] = None):
+        super().__init__(wires=wires, id=id)
+
```
```suggestion
```

dwierichs commented on pennylane/ops/qubit/non_parametric_ops.py:
```diff
@@ -1037,6 +1100,20 @@ class SX(Operation):
+            )
+        return self._pauli_rep
+
+    def __init__(self, wires: WiresLike, id: Optional[str] = None):
+        super().__init__(wires=wires, id=id)
+
```
```suggestion
```

dwierichs commented on pennylane/ops/qubit/non_parametric_ops.py:
```diff
@@ -1152,6 +1229,22 @@ class SWAP(Operation):
+        if self._pauli_rep is None:
+            self._pauli_rep = qml.pauli.PauliSentence(
+                {
+                    qml.pauli.PauliWord({self.wires[0]: "I", self.wires[1]: "I"}): 0.5,
```
Could do
```suggestion
                    qml.pauli.PauliWord({}): 0.5,
```
here, as empty slots in a `PauliWord` are the same as putting an identity in. However, the current version is nice and symmetric, so I'll leave it to you/reviewer #2 to decide on this one :)

dwierichs commented on tests/ops/qubit/test_non_parametric_ops.py:
```diff
@@ -66,6 +67,20 @@
 ]
 
+NON_PARAMETRIZED_OPERATIONS_WITH_PAULI_REP_ALREADY_IMPLEMENTED = [
+    (qml.X(0), X),
```
Also, you could maybe initialize some of the ops on non-default wires (some larger numbers, strings or any hashable are allowed as well, really).

ldi18 replied:
Okay, I added the comparisons with `gate_data`.

dwierichs commented on tests/ops/qubit/test_parametric_ops.py:
```diff
@@ -4049,3 +4060,23 @@ def test_op_aliases_are_valid():
+        """Compares the matrix representation obtained after using the .pauli_rep attribute with the result of the .matrix() method."""
+        import re
+
+        for op in SINGLE_QUBIT_PARAMETRIZED_OPERATIONS:
```
Same comment regarding using `pytest.mark.parametrize`

dwierichs reviewed (changes requested):
Thanks for this nice addition @ldi18 :tada: 
I had a couple of comments and proposed changes.

Nice catch regarding the PCPhase bug! If it is possible, it would be great to add a dedicated test for this bug. We know that it has been caught by the current test suite, but only in a very indirect manner. So it would be nice to check correctness of `PauliSentence.to_mat` or something like that directly in a small test.

ldi18 commented on pennylane/ops/qubit/parametric_ops_single_qubit.py:
```diff
@@ -74,11 +76,23 @@ class RX(Operation):
 
     def __init__(self, phi: TensorLike, wires: WiresLike, id: Optional[str] = None):
         super().__init__(phi, wires=wires, id=id)
+        self._phi = phi
```
Okay, nice!

ldi18 commented on tests/ops/qubit/test_parametric_ops.py:
```diff
@@ -4049,3 +4060,23 @@ def test_op_aliases_are_valid():
+        import re
+
+        for op in SINGLE_QUBIT_PARAMETRIZED_OPERATIONS:
+            dim = op.matrix().shape[0]
+            if dim == 2:
+                id_str = "I(0)"
+            if dim == 4:
+                id_str = "I(0)@I(1)"
+            op_in_pauli_decomp = str(op.pauli_rep).replace("\n", " ").replace("I", id_str)
+            for pauli_op in ["I", "X", "Y", "Z"]:
+                op_in_pauli_decomp = re.sub(
+                    rf"{pauli_op}", "qml." + rf"{pauli_op}", op_in_pauli_decomp
+                )
+            op_in_pauli_decomp = eval(op_in_pauli_decomp).matrix()  # pylint: disable=eval-used
```
That's much simpler. Great that it works.

ldi18 commented on tests/ops/qubit/test_non_parametric_ops.py:
```diff
@@ -1313,3 +1328,23 @@ def test_hadamard_alias(self):
+
+
+class TestPauliRep:
+    def test_matrix_and_pauli_rep_equivalence_for_all_non_parametric_ops(self):
```
Yes!

ldi18 commented:
@dwierichs Thanks for the feedback! Regarding the tensor-like parameter bug: I could implement shape-based test cases like
```python
assert qml.RX([0.1, 0.23, 1.11], wires=0).matrix().shape == (3, 2, 2)
````
Such test cases should then also include products of Pauli sentences like
```python
assert (qml.RX([1, 2, 3], wires=0) @ qml.RY([1, 2, 3], wires=0)).matrix().shape == (3, 2, 2)
```
The test case above actually fails unless I use 
```python
(qml.RX(np.array([1, 2, 3]), wires=0) @ qml.RY(np.array([1, 2, 3]), wires=0)).matrix().shape == (3, 2, 2)
```
It seems there might be a missing list-to-array conversion. Are lists not considered tensor-like in this context? Should I open a new issue for this and include more test cases for Pauli arithmetic in a PR linked to that issue?

dwierichs commented:
Hi @ldi18 
I'm actually unsure how much further we want dive into the whole batched-parameter + Pauli arithmetic topic, because it opens up a whole can of worms regarding batching support in `PauliSentence` and the way `qml.matrix` computes matrices.
Do we have any way to circumvent this problem in this PR without fixing the underlying bug in this PR?
We could, for example, skip adding the pauli rep to `PhaseShift`.
It's not a beautiful solution, but it allows us to disentangle the batching support from this PR.

cc @albi3ro @isaacdevlugt

ldi18 commented:
> Do we have any way to circumvent this problem in this PR without fixing the underlying bug in this PR? We could, for example, skip adding the pauli rep to `PhaseShift`. It's not a beautiful solution, but it allows us to disentangle the batching support from this PR.

@dwierichs Not sure if I get you right but the batching problem in pauli arithmetic applied to all parametrized gates with a `pauli_rep`. This can be seen from the test cases I added now, since they all fail when I revert the changes in `_sum_same_structure_pws_dense()`. 

```python
    def test_tensor_like_inputs(self):
        """Tests that the shape of (flattened) tensor like inputs is preserved in pauli arithmetic."""
        assert qml.RX([0.1, 0.23, 1.11], wires=0).matrix().shape == (3, 2, 2)
        assert (
            qml.U1([[0.1, 0.23, 1.11, 2.34], [1.1, 1.23, 2.11, 3.34], [3.1, 3.23, 3.11, 3.34], [1.1, 1.23, 1.11, 1.34]],
                wires="qubit0",
            ).matrix().shape)
        assert (qml.RX([1.11, 2.11, 3.11], wires=0) @ qml.RY([1.23, 2.23, 3.23], wires=0)).matrix().shape == (3, 2, 2)
        assert (qml.RX([1.11, 2.11, 3.11], wires=0) @ qml.RY([1.23, 2.23, 3.23], wires="qubit1")).matrix().shape == (3, 4, 4)
        assert (qml.RX([1.11, 2.11, 3.11], wires="qubit0") @ qml.RY([1.23, 2.23, 3.23], wires="qubit1") @ qml.RY([1.23, 2.23, 3.23], wires="qubit3")).matrix().shape == (3, 8, 8)
        assert (qml.IsingXX([1.11, 2.11, 3.11, 4.11, 5.11], wires=[1, 2]) @ qml.PhaseShift([1.23, 2.23, 3.23, 4.23, 5.23], wires=2) @ qml.PCPhase([1.23, 2.23, 3.23, 4.23, 4.23], dim=2, wires=1)
        ).matrix().shape == (5, 4, 4)
        assert (qml.Rot([1.11, 2.11, 3.11, 4.11, 5.11], [1.11, 2.11, 3.11, 4.11, 5.11], [1.11, 2.11, 3.11, 4.11, 5.11], wires=0) @ qml.Rot([1.11, 2.11, 3.11, 4.11, 5.11], [1.11, 2.11, 3.11, 4.11, 5.11], [1.11, 2.11, 3.11, 4.11, 5.11], wires=1)).matrix().shape == (5, 4, 4)
```


Actually I can't reproduce the error mentioned above from
```python
assert (qml.RX([1, 2, 3], wires=0) @ qml.RY([1, 2, 3], wires=0)).matrix().shape == (3, 2, 2)
```
so it was probably just a problem on my setup. Tensor-like parameters seem to work well as long as they are flattened. For example
```python
print(qml.RX([[0.1, 0.23, 1.11, 2.34], [1.1, 1.23, 2.11, 3.34], [3.1, 3.23, 3.11, 3.34], [1.1, 1.23, 1.11, 1.34]], wires="qubit0").matrix().shape)
print(qml.U1([[0.1, 0.23, 1.11, 2.34], [1.1, 1.23, 2.11, 3.34], [3.1, 3.23, 3.11, 3.34], [1.1, 1.23, 1.11, 1.34]], wires="qubit0").matrix().shape)
``` 
gives me different shapes - but that goes beyond this PR.

dwierichs commented:
Ah, I think we indeed usually do not allow lists for batched inputs. Sorry I had missed to answer this part.

I agree that the bug affects all gates, but without adding your test cases, skipping the `pauli_rep` of `PhaseShift` will allow us to proceed with this PR and do the batching-related things in a new PR ;-)

albi3ro commented on pennylane/ops/qubit/parametric_ops_single_qubit.py:
```diff
@@ -580,6 +628,26 @@ class Rot(Operation):
+                {
+                    qml.pauli.PauliWord({self.wires[0]: "I"}): qml.math.cos(self.data[1] / 2)
+                    * qml.math.cos((self.data[0] + self.data[2]) / 2),
+                    qml.pauli.PauliWord({self.wires[0]: "X"}): -1j
+                    * qml.math.sin(self.data[1] / 2)
+                    * qml.math.sin((self.data[0] - self.data[2]) / 2),
```
Given these are starting to get a big longer, can we pull them out into their own steps:
```
x_coeff = -1j * qml.math.sin(self.data[1]/2) * qml.math.sin((self.data[0] - self.data[2]) / 2)
...
qml.pauli.PauliWord({self.wires[0]: "X"}): x_coeff
```
should enhance readability a little bit.

albi3ro commented on pennylane/ops/qubit/parametric_ops_single_qubit.py:
🤞 this "shouldn't" cause circular import issues, but it might be worth doing `from pennylane.pauli import PauliWord, PauliSentence`.  But we do sometimes have those circular dependency issues.

albi3ro commented on pennylane/ops/qubit/parametric_ops_single_qubit.py:
```diff
@@ -756,6 +824,19 @@ class U1(Operation):
+        if self._pauli_rep is None:
+            self._pauli_rep = qml.pauli.PauliSentence(
+                {
+                    qml.pauli.PauliWord({self.wires[0]: "I"}): 0.5
```
```suggestion
                    qml.pauli.PauliWord({}): 0.5
```
Anything with `I` gets immediately pruned away, so might as well leave it out.

dwierichs commented on pennylane/ops/qubit/parametric_ops_single_qubit.py:
```diff
@@ -884,6 +962,27 @@ class U2(Operation):
+        if self._pauli_rep is None:
+            self._pauli_rep = qml.pauli.PauliSentence(
+                {
+                    qml.pauli.PauliWord({self.wires[0]: "I"}): 0.5
+                    * INV_SQRT2
+                    * (1 + qml.math.exp(1j * (self.data[1] + self.data[0]))),
+                    qml.pauli.PauliWord({self.wires[0]: "X"}): 0.5
+                    * INV_SQRT2
+                    * (qml.math.exp(1j * self.data[0]) - qml.math.exp(1j * self.data[1])),
+                    qml.pauli.PauliWord({self.wires[0]: "Y"}): -0.5j
+                    * INV_SQRT2
+                    * (qml.math.exp(1j * self.data[0]) + qml.math.exp(1j * self.data[1])),
+                    qml.pauli.PauliWord({self.wires[0]: "Z"}): 0.5
+                    * INV_SQRT2
+                    * (1 - qml.math.exp(1j * (self.data[1] + self.data[0]))),
```
This one I did not find suuuper readable, maybe again extract coefficient computation. A matter of taste, though.

dwierichs commented on pennylane/ops/qubit/parametric_ops_single_qubit.py:
```diff
@@ -1031,6 +1130,27 @@ class U3(Operation):
+        if self._pauli_rep is None:
+            self._pauli_rep = qml.pauli.PauliSentence(
+                {
+                    qml.pauli.PauliWord({self.wires[0]: "I"}): 0.5
+                    * (1 + qml.math.exp(1j * (self.data[2] + self.data[1])))
+                    * qml.math.cos(self.data[0] / 2),
+                    qml.pauli.PauliWord({self.wires[0]: "X"}): -0.5
+                    * (qml.math.exp(1j * self.data[2]) - qml.math.exp(1j * self.data[1]))
+                    * qml.math.sin(self.data[0] / 2),
+                    qml.pauli.PauliWord({self.wires[0]: "Y"}): -0.5j
+                    * (qml.math.exp(1j * self.data[2]) + qml.math.exp(1j * self.data[1]))
+                    * qml.math.sin(self.data[0] / 2),
+                    qml.pauli.PauliWord({self.wires[0]: "Z"}): 0.5
+                    * (1 - qml.math.exp(1j * (self.data[2] + self.data[1])))
+                    * qml.math.cos(self.data[0] / 2),
```
This one I did not find suuuper readable, maybe again extract coefficient computation. A matter of taste, though.

dwierichs commented on .gitignore:
```diff
@@ -31,3 +31,4 @@ datasets/*
 .benchmarks/*
 *.h5
 *.hdf5
+.venv
```
```suggestion
```

dwierichs reviewed (approved):
Thanks for this contribution, @ldi18 , looks good to me!
And thanks for your patience.

Pylint seems to have some comments (see CI below).

JerryChen97 reviewed (dismissed):
LGTM! Just some format related comments hopefully helpful

dwierichs commented on tests/ops/qubit/test_parametric_ops.py:
```diff
@@ -20,6 +20,7 @@
 
 import numpy as np
 import pytest
+import torch as torch
```
`torch` is not  readily available in the test suite. I think it is fine to not go for interface tests here.
```suggestion
```

dwierichs commented on tests/ops/qubit/test_parametric_ops.py:
```diff
@@ -68,6 +69,25 @@
+    qml.RZ(2.774, wires=-1),
+    qml.Rot(0.123, 0.456, 0.789, wires="a"),
+    qml.U1(0.123, wires="qubit1"),
+    qml.U2(
+        torch.tensor(3.556),
+        torch.tensor(2.134),
+        wires="qubit2",
+    ),
```
```suggestion
    qml.U2(3.556, 2.134, wires="qubit2"),
```

dwierichs commented:
@ldi18 I took the liberty to incorporate @JerryChen97 's suggestions, and to resolve a small annoyance regarding `torch` in the test suite (removed it because we don't need to test that combination necessarily, and it would require setting up an independent test).
I hope that's all fine by you.

Thanks for the nice contribution, and sorry about the holdup on this one! 
Let me know whether the PR is good to merge from your end :)

JerryChen97 commented:
My apology for being too hasty to go over and approve. Didn't noticed that the passed ci tests are all night bumps instead of full checks. I dismiss my approval for now

JerryChen97 commented:
Besides the interface specific tests, I guess the [core test batch 5](https://github.com/PennyLaneAI/pennylane/actions/runs/12039957509/job/33568905979?pr=6562) contains more information. I read the log and here's a lazy sum:

## 1. Pauli Representation Extension Issues
### Tests:
- `test_pauli_rep_none_if_base_pauli_rep_none`
- `test_terms_no_pauli_rep`
- `test_pauli_rep_none`
- `test_non_pauli_word_ops_not_supported`
- `test_grouping_raises_error`
- `test_non_pauli_error`
- `test_pauli_raises_NotImplementedError_without_pauli_rep`
- `test_measurement_expansion`
- `test_are_qwc_pauli_words`
- `test_error_not_linear_comb_pauli_words`

> Consider: These tests were designed with assumption of limited Pauli rep support; need update to match new feature scope

## 2. Operation Type & Simplification Issues
### Tests:
- `test_simplify_method` 
- `test_simplify_adj_of_prod`
- `test_simplify_method_with_nested_ops`
- `test_simplify_grouping`
- `test_simplify_with_pow_not_defined`
- `test_simplify_with_nested_prod_and_adjoints`

> Consider: Review type preservation in simplification chain

## 3. Matrix/Broadcasting Problems
### Tests:
- `test_matrix_all_batched`
- `test_matrix_not_all_batched`
- `test_batched_operation`
- `test_terms_mixed`
- `test_simplify_grouping_with_tolerance`

> Consider: Add consistent shape validation pre-operation

## 4. Integration & High-Level Operations
### Tests:
- `test_measurement_expansion`
- `test_approx_time_evolution`
- `test_adjoint_repr`
- `test_structure_constants`

> Consider: Review integration points with extended Pauli reps

dwierichs commented:
Hi @ldi18 ,
Unfortunately, more problems with this feature addition were surfaced by the CI tests here.
We discussed internally and decided that a reduced version of the addition that does not cause these problems would be the best way to proceed. In particular, the reduction is to only add `pauli_rep` for the non-parametrized gates, and not to `Hadamard`. I tried this out in #6646 . 

Please let me know whether you would adapt your PR to this reduction yourself or whether I should take care of this. :) 

Sorry again for the inconveniences!

ldi18 commented:
Hi @dwierichs, 
Thanks for the help. I deleted all the pauli reps for the parametrized ops and the Hadamard gate from this PR. I also removed the corresponding tests.

dwierichs reviewed (approved):
Thanks for the update to the new version @ldi18 Looks good to me! :100: 

For some reason, this PR is cursed and now has a technical problem in a Github action after testing the documentation creation. I'll follow up internally on this.

andrijapau commented on .gitignore:
```diff
@@ -30,4 +30,4 @@ qml_debug.log
 .benchmarks/*
 *.h5
-*.hdf5
+*.hdf5
```
I think files need to end with a newline for pylint to be happy. 😄 
```suggestion
*.hdf5

```

andrijapau reviewed (approved):
Minor code suggestion to make pylint happy, but otherwise it LGTM! Thanks for all your help! 🏅

dwierichs commented:
Just waiting for #6661 to be merged, and then this PR should be merge-ready :)
